### DISCUSSION TITLE
axios download for laravel generated excel files

### DISCUSSION THREAD

**Discussion by kndrckjvr** [spatie/simple-excel#169] [Q&A]
We wanted to download the excel file using an authenticated route to have a check of permissions before downloading.

This is the code for my Controller and trying just with a sample one. but I am still having problems with CORS. I have tried applying a CORS Middleware from different sources but still doesn't work.

All of my other routes have CORS Headers applied to them.

I don't have any answers why it is not working. I have tried also applying the headers from the response header parameter as seen on the sample code and yet the response is still not applying the CORS headers

Sample Code:
```
$data = SimpleExcelWriter::streamDownload('sample.xlsx')
    ->addHeader(['first_name', 'last_name'])
    ->addRow(['John', 'Doe'])
    ->addRow(['Jane', 'Doe']);

return response()->streamDownload(fn () => $data->close(), 'sample.xlsx', [
    'Access-Control-Allow-Origin' => '*',
    'Access-Control-Allow-Methods' => 'POST, GET, OPTIONS, PUT, DELETE',
    'Access-Control-Allow-Credentials' => 'true',
    'Access-Control-Max-Age' => '86400',
    'Access-Control-Allow-Headers' => 'Content-Type, Accept, Authorization, X-Requested-With, Application',
]);
```

Also tried working with just returning the whole SimpleExcelWriter but still having the CORS error:

```
return SimpleExcelWriter::streamDownload('sample.xlsx')
    ->addHeader(['first_name', 'last_name'])
    ->addRow(['John', 'Doe'])
    ->addRow(['Jane', 'Doe'])
    ->toBrowser();
```

**Comment by thaihacong** [comment]
Based on https://github.com/spatie/simple-excel/discussions/149#discussioncomment-9298723, I got it to work by adding
```
    // fix error: Access to XMLHttpRequest ... has been blocked by CORS policy ('cors'):
    // No 'Access-Control-Allow-Origin' header is present on the requested resource.
    header('Access-Control-Allow-Origin: *');

    // extra
    // fix missing some other response headers, e.g. 'content-disposition' to get export file name
    header('Access-Control-Expose-Headers: *');
```
before calling `SimpleExcelWriter::streamDownload('sample.xlsx')`.
